Python binding to set/get CUDA rng state offset

anijain2305 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #98988
* __->__ #98965


Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~

anijain2305 committed:
Python binding to set CUDA rng state offset

[ghstack-poisoned]

albanD commented on aten/src/ATen/CPUGeneratorImpl.cpp:
```diff
@@ -94,6 +94,14 @@ void CPUGeneratorImpl::set_current_seed(uint64_t seed) {
+ * See Note [Acquire lock when using random generators]
+ */
+void CPUGeneratorImpl::set_offset(uint64_t offset) {
+  AT_ERROR("CPU Generator does not use offset");
```
TORCH_CHECK

anijain2305 replied:
What do you mean TORCH_CHECK? Here, my intent is - "you should not be here".
Earlier I was thinking - `throw std::runtime_error("not implemented");`

ngimel replied:
`TORCH_CHECK` is something we use when user provides invalid input, we should deprecate AT_ERROR and use TORCH_CHECK in its place

anijain2305 replied:
Ah, I see. Something like `TORCH_CHECK(false, ...)`. Sounds good.

albanD replied:
Yes AT_ERROR is a deprecated macro. TORCH_CHECK and TORCH_INTERNAL_ASSERT should be used.

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.


[ghstack-poisoned]

albanD commented on aten/src/ATen/mps/MPSGeneratorImpl.mm:
```diff
@@ -37,6 +37,10 @@ Generator createMPSGenerator(uint64_t seed_val) {
   engine_.reset_state(seed);
 }
 
+void MPSGeneratorImpl::set_offset(uint64_t offset) {
```
I'm sure we have an offset here no?
cc @kulinseth

anijain2305 replied:
I read the code, but it was not clear to me.

kulinseth replied:
[Here](https://github.com/pytorch/pytorch/pull/98965/files#diff-3a29e57c9df0293adc5940e0bc806ad0252d1c8f9303e97112eeb35603e4fb8bR59) we update the Philox state. The use is as follows: 
1. We start with the initialized state 
2. Pass it to our random op, which returns the updated Philox state
3. And then we write out the [state.](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/mps/operations/Distributions.mm#L106)

We have a state maintained [here](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/mps/MPSGeneratorImpl.h#L16). You can add offset parameter to update the state.

albanD commented on aten/src/ATen/core/Generator.h:
```diff
@@ -93,6 +93,7 @@ struct TORCH_API Generator {
   }
 
   void set_current_seed(uint64_t seed) { impl_->set_current_seed(seed); }
+  void set_offset(uint64_t offset) { impl_->set_offset(offset); }
```
We should really define what this does!

albanD commented on aten/src/ATen/CPUGeneratorImpl.cpp:
```diff
@@ -94,6 +94,14 @@ void CPUGeneratorImpl::set_current_seed(uint64_t seed) {
+ * See Note [Acquire lock when using random generators]
+ */
+void CPUGeneratorImpl::set_offset(uint64_t offset) {
+  AT_ERROR("CPU Generator does not use offset");
```
Also offset of 0 sounds ok.

albanD commented on torch/csrc/Generator.cpp:
```diff
@@ -150,6 +150,40 @@ static PyObject* THPGenerator_manualSeed(PyObject* _self, PyObject* seed) {
+      "but got %s",
+      THPUtils_typename(offset));
+  // See Note [Acquire lock when using random generators]
+  std::lock_guard<std::mutex> lock(generator.mutex());
```
Don't you only need that lock when you do the set on the generator below?

albanD commented on torch/csrc/Generator.cpp:
```diff
@@ -150,6 +150,40 @@ static PyObject* THPGenerator_manualSeed(PyObject* _self, PyObject* seed) {
+  }
+  generator.set_offset(offset_unpacked);
+  Py_INCREF(self);
+  return (PyObject*)self;
```
you can just `return _self;`

albanD commented on torch/cuda/random.py:
```diff
@@ -64,6 +65,30 @@ def cb():
+    r"""Sets the random number generator state offset of the specified GPU.
+
+    Args:
+        new_state (torch.ByteTensor): The desired state
```
offset

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

ngimel commented on torch/csrc/Generator.cpp:
```diff
@@ -144,9 +142,45 @@ static PyObject* THPGenerator_manualSeed(PyObject* _self, PyObject* seed) {
+  } catch (...) {
+    if (PyErr_ExceptionMatches(PyExc_OverflowError)) {
+      // If an overflow happened, then the offset could be negative,
+      // so try to interpret it as signed long
```
does philox generator accept negative offsets?

anijain2305 replied:
Good point. I basically copied this from looking at manual seed implementation. seed can be a negative value. 

But negative offset does not make sense to me. 

cc @albanD if you also agree, I will remove the overflow handling part.

albanD replied:
Well, both the seed and the offset should be `uint64_t`.
Python integer are infinite precision though so we do something weird to allow negative values.

I do agree that it doesn't matter much though, we could just pass in `n % UINT64_MAX` (the version that always returns a positive number) as well to allow any python integer value in a nicely consistent way.

anijain2305 replied:
Ok, I am not changing it then. Keeping it same as `seed` to keep it consistent.

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 commented:
cc @kulinseth Can you please take a look again? I added support for offset in MPS PhiloxState.

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

kulinseth reviewed (approved):
Looks good to me.

ezyang commented on aten/src/ATen/core/Generator.h:
```diff
@@ -93,6 +93,9 @@ struct TORCH_API Generator {
   void set_current_seed(uint64_t seed) { impl_->set_current_seed(seed); }
+  // Sets the offset of Generator state to the desired offset. This is currently
+  // supported for only Philox based Generators, i.e., CUDA and MPS.
+  void set_offset(uint64_t offset) { impl_->set_offset(offset); }
```
It's a bit annoying that you are putting a subclass specific method un the base class

anijain2305 commented:
@pytorchbot merge

ezyang commented on torch/cuda/random.py:
```diff
@@ -64,6 +65,30 @@ def cb():
+        idx = cast(torch.device, device).index
+        if idx is None:
+            idx = current_device()
+        default_generator = torch.cuda.default_generators[idx]
```
This is very copy-pastey with the function above. Refactor?

anijain2305 replied:
Done

voznesenskym commented:
Tests??

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"

Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())

tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~

[ghstack-poisoned]

anijain2305 commented:
> Tests??

Added one. Thanks for asking!

voznesenskym commented on torch/csrc/Generator.cpp:
```diff
@@ -116,37 +116,60 @@ static PyObject* THPGenerator_setState(PyObject* _self, PyObject* _new_state) {
+uint64_t unpack_int_pyobj(PyObject* obj) {
   // NOLINTNEXTLINE(cppcoreguidelines-init-variables)
-  uint64_t seed_unpacked;
+  uint64_t obj_unpacked;
```
nitpick: The "unsigned" being default named as obj_unpacked, and the signed being named "obj_unpacked_signed" is kinda backwards from the c++ convention of unsigned being named and signed not being named.

Nitpick #2: Try-catch for overflow like this can be improved to use PyError

```
uint64_t obj_unpacked;
if (PyLong_Check(obj)) {
  uint64_t value = PyLong_AsUnsignedLongLong(obj);
  if (value == (uint64_t)-1 && PyErr_Occurred()) {
    PyErr_Clear();
    obj_unpacked = *(reinterpret_cast<uint64_t*>(&THPUtils_unpackLong(obj);));
  } else {
    obj_unpacked = value;
  }
} else {
  obj_unpacked = THPUtils_unpackUInt64(obj);
}
```

voznesenskym commented on torch/csrc/Generator.cpp:
```diff
@@ -116,37 +116,60 @@ static PyObject* THPGenerator_setState(PyObject* _self, PyObject* _new_state) {
+      "but got %s",
+      THPUtils_typename(seed));
+  uint64_t seed_unpacked = unpack_int_pyobj(seed);
+  // See Note [Acquire lock when using random generators]
```
I wonder if you can do thread-safe generators by making thread affine/thread local copies of underlying data. cc @albanD

voznesenskym commented on torch/csrc/Generator.cpp:
```diff
@@ -116,37 +116,60 @@ static PyObject* THPGenerator_setState(PyObject* _self, PyObject* _new_state) {
+      THPUtils_typename(seed));
+  uint64_t seed_unpacked = unpack_int_pyobj(seed);
+  // See Note [Acquire lock when using random generators]
+  std::lock_guard<std::mutex> lock(generator.mutex());
```
lock_guard c++17 deprecated, I dunno what version we are on, but if its 17 or higher, std::scoped_lock

anijain2305 replied:
Thanks for pointing this out. `CMAKE_CXX_STANDARD` is set to C17, so I think we are building with c++17. However, currently almost all of PyTorch uses `lock_guard`. So, lets do this in a separate PR. opened an issue for this - https://github.com/pytorch/pytorch/issues/99359

voznesenskym replied:
wait, why? Just do the right thing here.

ezyang commented on torch/cuda/random.py:
```diff
@@ -49,21 +44,33 @@ def set_rng_state(new_state: Tensor, device: Union[int, str, torch.device] = 'cu
     _lazy_call(cb)
 
 
+def set_rng_state_offset(offset: int, device: Union[int, str, torch.device] = 'cuda') -> None:
```
You're publishing this as public API but it's not entirely clear this makes sense as public API? Some questions that would at least have to be documented here: When do I use `set_rng_state_offset` versus `set_rng_state`? What's the difference? Can I use these two APIs together? Do I get deterministic results if I reset two distinct RNGs to the same offset?

If we you don't want to upfront answer these questions, it would seem better to NOT make this public API to start.

voznesenskym commented on torch/csrc/Generator.cpp:
```diff
@@ -116,37 +116,60 @@ static PyObject* THPGenerator_setState(PyObject* _self, PyObject* _new_state) {
+  // See Note [Acquire lock when using random generators]
+  std::lock_guard<std::mutex> lock(generator.mutex());
+  generator.set_offset(offset_unpacked);
+  Py_INCREF(self);
```
1) Incref where you get the object, not where you return it, so move this up to `auto self`
2) This returns a borrowed ref to _self, as it has an incref. Where is it cleared? It is a little odd to me that setting the offset incurs a reference count. I know THPGenerator_setState does this too, but I don't get it. Can you explain the lifecycle here please?

anijain2305 replied:
@albanD helped us understand this. Summary
* Arguments are always borrowed. So no need to incref them.
* CPython calling convention says that we need to send new references. Therefore we incref the new reference just before the return. Now CPython owns this reference, and will call DECREF when the new reference goes out of scope.

voznesenskym commented on aten/src/ATen/mps/MPSGeneratorImpl.mm:
```diff
@@ -29,12 +29,19 @@ Generator createMPSGenerator(uint64_t seed_val) {
 
 void MPSGeneratorImpl::set_current_seed(uint64_t seed) {
   data_.seed = seed;
+  data_.offset = 0;
```
this is redundant with set_offset right?

voznesenskym commented on test/test_cuda.py:
```diff
@@ -1699,6 +1699,15 @@ def test_get_set_rng_state_all(self):
         self.assertEqual(before0, after0, atol=0, rtol=0)
         self.assertEqual(before1, after1, atol=0, rtol=0)
 
+    @unittest.skipIf(not TEST_MULTIGPU, "only one GPU detected")
```
Thank you for adding this :)

ezyang commented on aten/src/ATen/mps/MPSGeneratorImpl.mm:
```diff
@@ -29,12 +29,19 @@ Generator createMPSGenerator(uint64_t seed_val) {
 
 void MPSGeneratorImpl::set_current_seed(uint64_t seed) {
   data_.seed = seed;
+  data_.offset = 0;
```
This is a strange thing to do. Why add offset to MPS generator if you're not actually going to use it for anything. 🤔

ezyang replied:
OK, so it looks like `offset` is not used for anything, but engine does use it. But then that begs the question, why store the offset in two places?

anijain2305 replied:
This is due is half laziness and other half that I did not want to change the state handling too much (mostly the latter one). `seed` gets the similar treatment, and we store it at 2 places. Both seed and offset are used in get_rng_state and set_rng_state.

But this is not a good answer. Let me see if I can clean this up better.

anijain2305 replied:
I fixed this in latest commit. Offset state is managed inside the Philox Engine now.

voznesenskym commented on torch/cuda/__init__.py:
```diff
@@ -352,6 +352,33 @@ def __init__(self, obj):
 
 
+
+def _get_device(device: Union[int, str, torch.device] = 'cuda') -> torch.device:
```
Why does this have a default?

ngimel replied:
It's used as a utility function in many functions that omit device arg at callsite, e.g. `torch.cuda.get_rng_state()` is supposed to return rng state on current device.

voznesenskym commented on torch/cuda/__init__.py:
```diff
@@ -352,6 +352,33 @@ def __init__(self, obj):
+    Args:
+        device (torch.device or int): selected device.
+    """
+    if isinstance(device, str):
+        device = torch.device(device)
+    elif isinstance(device, int):
+        device = torch.device('cuda', device)
+    return device
```
This just passes through if its neither of these types? Add an assert.

voznesenskym commented on torch/cuda/__init__.py:
```diff
@@ -352,6 +352,33 @@ def __init__(self, obj):
+    idx = device.index
+    if idx is None:
+        idx = current_device()
+    return torch.cuda.default_generators[idx]
```
wait what if its a CPU device? is this sound? Do you need to assert that this is cuda?

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "Python binding to set CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "Python binding to set/get CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "Python binding to set/get CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "Python binding to set/get CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "Python binding to set/get CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

ezyang commented:
this is my main blocker to approving: https://github.com/pytorch/pytorch/pull/98965#discussion_r1168009188

anijain2305 committed:
Update on "Python binding to set/get CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 committed:
Update on "Python binding to set/get CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 commented:
@ezyang @voznesenskym Thanks for being super patient with the reviews. I have incorporated your comments. Please review again.

ezyang commented on aten/src/ATen/cuda/CUDAGeneratorImpl.cpp:
```diff
@@ -117,6 +117,27 @@ void CUDAGeneratorImpl::set_current_seed(uint64_t seed) {
+uint64_t CUDAGeneratorImpl::get_offset() const {
+  // Debatable if get_offset() should be allowed in captured regions.
+  // Conservatively disallow it for now.
+  at::cuda::assertNotCapturing("Cannot call CUDAGeneratorImpl::current_seed");
```
typo

anijain2305 committed:
Update on "Python binding to set/get CUDA rng state offset"



Why?
* To reduce the latency of hot path in https://github.com/pytorch/pytorch/pull/97377

Concern - I had to add `set_offset` in all instances of `GeneratorImpl`. I don't know if there is a better way.

~~~~
import torch
torch.cuda.manual_seed(123)
print(torch.cuda.get_rng_state())
torch.cuda.set_rng_state_offset(40)
print(torch.cuda.get_rng_state())


tensor([123,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
tensor([123,   0,   0,   0,   0,   0,   0,   0,  40,   0,   0,   0,   0,   0,
          0,   0], dtype=torch.uint8)
~~~


[ghstack-poisoned]

anijain2305 commented:
@pytorchbot merge

anijain2305 commented:
@pytorchbot merge

DanilBaibak commented:
@pytorchbot revert -m "Break internal build" -c ghfirst

DanilBaibak commented:
@anijain2305 sorry, I have to revert your PR. It leads to internal failures. Here you can find more details - [D45103441](https://www.internalfb.com/diff/D45103441).

malfet commented:
Sorry, for the revert, looks like all the failures are in integration with https://github.com/pytorch/csprng which is dead.

anijain2305 commented:
@malfet Whats the next step here? Can I reland this if pytorch/csprng is dead or do we need something else?
